### PR TITLE
HEEDLS-861 - add additional pagination controls at top of lists

### PR DESCRIPTION
**Pull request by DanBloxham-sw**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-861

### Description
Created a new pagination configuration that can be placed at the top of lists to provide an additional set of pagination controls. I've had to refactor the pagination JS/views to allow the set up to work with multiple sets of pagination controls. I decided to split the pagination controls into two sets of partials, rather than just a single configurable one, as I thought the end product would be a bit neater to work with versus having a bunch of tuples everywhere. If you can think of a nicer solution, I'll be happy to use it.

Changed the ellipsis in the loading page warning to a full stop.

Mysteriously, the announcement of page numbers and result count seems to work as we would hope without any modifications (at least in Chrome). Even on the test site. So I have no changed anything for that, and we will have to see if we have any problems when this is tested.

I've also fixed a couple of bugs that are mentioned in https://softwiretech.atlassian.net/browse/HEEDLS-881

### Screenshots
Example page of All Delegates

Desktop:
All delegates desktop

Tablet

Mobile

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web/Views/Shared/PaginatedPage/_BaseConfigurablePaginationNav.cshtml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            <path d="M4.1 12.3l2.7 3c.2.2.5.2.7 0 .1-.1.1-.2.1-.3v-2h11c.6 0 1-.4 1-1s-.4-1-1-1h-11V9c0-.2-.1-.4-.3-.5h-.2c-.1 0-.3.1-.4.2l-2.7 3c0 .2 0 .4.1.6z"></path>
+          </svg>
+          <span class="nhsuk-pagination__title">Previous</span>
+        </button>
+      </div>
+    </div>
+    <div class="page-indicator-container">
+      <span @(Model.useAlertOnPageNumber ? "role=alert" : null)>
+        <span class="nhsuk-u-visually-hidden">page </span>
+        <span class="page-indicator" @(Model.pageModel.TotalPages <= 1 ? "hidden" : "")>@Model.pageModel.Page of @Model.pageModel.TotalPages</span>
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Views/Shared/PaginatedPage/_BaseConfigurablePaginationNav.cshtml:18]: I've added the logic to hide this when there is 1 or 0 pages to fix the bug from 881.

File: DigitalLearningSolutions.Web/Views/Shared/SearchablePage/_PaginationInputs.cshtml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿@using DigitalLearningSolutions.Web.ViewModels.Common.SearchablePage
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/Shared/SearchablePage/_PaginationInputs.cshtml:1]: I think this probably needs a different name, this implies to me it is the inputs that are required for the pagination functionality, not the other inputs for the rest of the stuff.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Changes overall look fine to me, just one issue with naming.

I also wonder if we should try to decrease this spacing on the top controls?
image

File: DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/paginate.ts [hunk 3]
```diff
@@ -42,31 +44,37 @@ export function paginateResults(
...
+  const nextButtons = getNextButtonDisplayContainers();
+  const paginationContainers = getPaginationDisplayContainers();
+
+  nextButtons.forEach((button) => {
+    const element = button;
+    element.hidden = page >= totalPages;
+  });
+  previousButtons.forEach((button) => {
+    const element = button;
+    element.hidden = page === 1;
```
Comment (GeorgeBlackwell) [inline DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/paginate.ts:71]: You can inline this and remove the need for `element`

**Review by GeorgeBlackwell — CHANGES_REQUESTED** [review]
Nice work, just a minor comment

**Comment by DanBloxham-sw** [comment]
After the Show and Tell today (14th April 2022) it was decided that the current spacing was fine, so I'll not change any of that around.

**Review by GeorgeBlackwell — APPROVED** [review]

**Pull request merged**